File: src/spec_decode/speculative.py
"""Speculative decoding.

The idea in one line: a cheap **draft** model proposes several tokens; the
expensive **target** model checks them all in a single forward pass; and a
carefully designed accept/reject rule guarantees the tokens that come out are
distributed *exactly* as if they had been sampled from the target one at a time.
You get the target's quality at a fraction of its serial cost — with no
approximation.

This module implements the sampler from Leviathan et al., 2023 ("Fast Inference
from Transformers via Speculative Decoding") and Chen et al., 2023. The
correctness claim is not a slogan: ``proof.induced_next_token_distribution``
computes, in closed form, the distribution this sampler actually emits and the
test suite asserts it equals the target's — for arbitrary distributions.

Everything here is deterministic given a seeded RNG, so results are reproducible
and every branch is testable without a GPU.
"""

from __future__ import annotations

from dataclasses import dataclass, field

import numpy as np

from .models import Model, _normalize, sample


def residual(p: np.ndarray, q: np.ndarray) -> np.ndarray:
    """The adjusted distribution to sample from after a rejection.

    When the draft's proposal is rejected we must *not* simply resample from the
    target, or the math breaks. We sample from the normalized positive part of
    ``p - q`` — the probability mass the target wanted that the draft under-
    supplied. This is the term that makes the whole scheme exact.
    """
    return _normalize(np.maximum(p - q, 0.0))


def accept_probability(p_t: float, q_t: float) -> float:
    """Probability of accepting a proposed token with target prob ``p_t`` and
    draft prob ``q_t``: ``min(1, p_t / q_t)``.

    A token the target likes at least as much as the draft is always accepted;
    one the draft over-proposes is accepted proportionally less often. ``q_t``
    is positive whenever the draft actually proposed the token.
    """
    if q_t <= 0.0:
        return 1.0
    return min(1.0, p_t / q_t)


@dataclass
class Trace:
    """Bookkeeping for one ``generate`` call — the raw material for measuring
    speedup honestly."""

    tokens: list[int] = field(default_factory=list)
    steps: int = 0
    target_passes: int = 0
    draft_passes: int = 0
    proposed: int = 0
    accepted: int = 0

    @property
    def acceptance_rate(self) -> float:
        return self.accepted / self.proposed if self.proposed else 0.0

    @property
    def tokens_per_step(self) -> float:
        return len(self.tokens) / self.steps if self.steps else 0.0


def speculative_step(
    target: Model,
    draft: Model,
    prefix: list[int],
    k: int,
    rng: np.random.Generator,
    trace: Trace | None = None,
) -> list[int]:
    """Run one speculative step and return the tokens it emits (between 1 and
    ``k + 1`` of them).

    1. The draft autoregressively proposes ``k`` tokens, remembering the draft
       distribution used at each position.
    2. The target scores all ``k`` proposal positions plus one look-ahead
       position — in a real model this is a single batched forward pass; here it
       is ``k + 1`` cheap distribution lookups, counted as **one** target pass.
    3. Each proposal is accepted or rejected in order. The first rejection stops
       the step and emits a residual-sampled token in place of the rejected one.
    4. If every proposal is accepted, a bonus token is sampled from the target's
       look-ahead distribution — so a fully-accepted step of size ``k`` emits
       ``k + 1`` tokens for the price of one target pass.
    """
    # 1. Draft proposes k tokens.
    proposals: list[int] = []
    q_dists: list[np.ndarray] = []
    ctx = list(prefix)
    for _ in range(k):
        q = draft.next_dist(ctx)
        x = sample(q, rng)
        proposals.append(x)
        q_dists.append(q)
        ctx.append(x)
    if trace is not None:
        trace.draft_passes += k
        trace.proposed += k

    # 2. Target scores every proposal position and one look-ahead, in one pass.
    p_dists: list[np.ndarray] = []
    ctx = list(prefix)
    for i in range(k):
        p_dists.append(target.next_dist(ctx))
        ctx.append(proposals[i])
    p_lookahead = target.next_dist(ctx)  # position k+1
    if trace is not None:
        trace.target_passes += 1
        trace.steps += 1

    # 3. Accept/reject in order.
    emitted: list[int] = []
    for i in range(k):
        x = proposals[i]
        a = accept_probability(float(p_dists[i][x]), float(q_dists[i][x]))
        if rng.random() < a:
            emitted.append(x)
            if trace is not None:
                trace.accepted += 1
        else:
            # Rejected: emit a residual sample and stop this step.
            emitted.append(sample(residual(p_dists[i], q_dists[i]), rng))
            if trace is not None:
                trace.tokens.extend(emitted)
            return emitted

    # 4. All accepted: free bonus token from the look-ahead distribution.
    emitted.append(sample(p_lookahead, rng))
    if trace is not None:
        trace.tokens.extend(emitted)
    return emitted


def generate_speculative(
    target: Model,
    draft: Model,
    prefix: list[int],
    n_tokens: int,
    k: int,
    rng: np.random.Generator,
) -> Trace:
    """Generate ``n_tokens`` via speculative decoding, returning a full
    :class:`Trace` (tokens plus the pass/acceptance counters used for speedup)."""
    if k < 1:
        raise ValueError("k must be >= 1")
    trace = Trace()
    ctx = list(prefix)
    while len(trace.tokens) < n_tokens:
        emitted = speculative_step(target, draft, ctx, k, rng, trace)
        ctx.extend(emitted)
    # A step can overshoot the requested count; trim so the output is exact.
    if len(trace.tokens) > n_tokens:
        trace.tokens = trace.tokens[:n_tokens]
    return trace


def generate_target_only(
    target: Model,
    prefix: list[int],
    n_tokens: int,
    rng: np.random.Generator,
) -> Trace:
    """The baseline: sample ``n_tokens`` from the target one at a time. Costs one
    target pass per token — the serial cost speculative decoding beats."""
    trace = Trace()
    ctx = list(prefix)
    for _ in range(n_tokens):
        p = target.next_dist(ctx)
        x = sample(p, rng)
        trace.tokens.append(x)
        trace.target_passes += 1
        trace.steps += 1
        ctx.append(x)
    return trace


def expected_speedup(tokens_per_step: float, k: int, draft_cost_ratio: float) -> float:
    """Wall-clock speedup implied by an observed ``tokens_per_step``.

    Per step the cost is one target pass plus ``k`` draft passes, where
    ``draft_cost_ratio`` = draft-pass-cost / target-pass-cost. Target-only would
    spend one target pass per emitted token. So::

        speedup = tokens_per_step / (1 + k * draft_cost_ratio)

    With a much cheaper draft (small ratio) the speedup approaches the average
    number of tokens emitted per step, capped at ``k + 1``.
    """
    return tokens_per_step / (1.0 + k * draft_cost_ratio)

File: java/src/main/java/com/specdecode/Speculative.java
package com.specdecode;

import java.util.ArrayList;
import java.util.List;

/**
 * The speculative-decoding sampler. Every routine is parameterised over an
 * {@link Rng} so the accept/reject state machine is deterministic.
 */
public final class Speculative {
    private Speculative() {
    }

    /**
     * Returns the normalised positive part of {@code p - q}: the residual
     * distribution sampled when a draft token is rejected.
     *
     * @param p the target distribution
     * @param q the draft distribution
     * @return the residual distribution
     */
    public static double[] residual(double[] p, double[] q) {
        double[] diff = new double[p.length];
        for (int i = 0; i < diff.length; i++) {
            double d = p[i] - q[i];
            diff[i] = d > 0.0 ? d : 0.0;
        }
        return Distributions.normalize(diff);
    }

    /**
     * Returns the probability of accepting a draft token whose target and draft
     * probabilities are {@code pt} and {@code qt}.
     *
     * @param pt the target probability of the token
     * @param qt the draft probability of the token
     * @return the acceptance probability
     */
    public static double acceptProbability(double pt, double qt) {
        if (qt <= 0.0) {
            return 1.0;
        }
        return Math.min(1.0, pt / qt);
    }

    private static int[] toArray(List<Integer> values) {
        int[] out = new int[values.size()];
        for (int i = 0; i < out.length; i++) {
            out[i] = values.get(i);
        }
        return out;
    }

    /**
     * Runs one speculative step: propose {@code k} draft tokens, score them
     * against the target, and emit the accepted prefix plus one bonus/residual
     * token. Emits at most {@code k + 1} tokens.
     *
     * @param target the target model
     * @param draft the draft model
     * @param prefix the current context
     * @param k the number of speculative proposals
     * @param rng the uniform source
     * @param trace optional bookkeeping (may be {@code null})
     * @return the emitted tokens
     */
    public static List<Integer> speculativeStep(
            Model target, Model draft, int[] prefix, int k, Rng rng, Trace trace) {
        List<Integer> proposals = new ArrayList<>(k);
        List<double[]> qDists = new ArrayList<>(k);
        List<Integer> ctx = new ArrayList<>();
        for (int t : prefix) {
            ctx.add(t);
        }
        for (int i = 0; i < k; i++) {
            double[] q = draft.nextDist(toArray(ctx));
            int x = Distributions.sample(q, rng);
            proposals.add(x);
            qDists.add(q);
            ctx.add(x);
        }
        if (trace != null) {
            trace.draftPasses += k;
            trace.proposed += k;
        }

        List<double[]> pDists = new ArrayList<>(k);
        ctx = new ArrayList<>();
        for (int t : prefix) {
            ctx.add(t);
        }
        for (int i = 0; i < k; i++) {
            pDists.add(target.nextDist(toArray(ctx)));
            ctx.add(proposals.get(i));
        }
        double[] pLookahead = target.nextDist(toArray(ctx));
        if (trace != null) {
            trace.targetPasses += 1;
            trace.steps += 1;
        }

        List<Integer> emitted = new ArrayList<>(k + 1);
        for (int i = 0; i < k; i++) {
            int x = proposals.get(i);
            double a = acceptProbability(pDists.get(i)[x], qDists.get(i)[x]);
            if (rng.nextUniform() < a) {
                emitted.add(x);
                if (trace != null) {
                    trace.accepted += 1;
                }
            } else {
                emitted.add(Distributions.sample(residual(pDists.get(i), qDists.get(i)), rng));
                if (trace != null) {
                    trace.tokens.addAll(emitted);
                }
                return emitted;
            }
        }
        emitted.add(Distributions.sample(pLookahead, rng));
        if (trace != null) {
            trace.tokens.addAll(emitted);
        }
        return emitted;
    }

    /**
     * Generates at least {@code nTokens} tokens via repeated speculative steps,
     * trimming any overshoot from the final step.
     *
     * @param target the target model
     * @param draft the draft model
     * @param prefix the initial context
     * @param nTokens the number of tokens to generate
     * @param k the number of speculative proposals per step
     * @param rng the uniform source
     * @return the completed trace
     * @throws IllegalArgumentException if {@code k < 1}
     */
    public static Trace generateSpeculative(
            Model target, Model draft, int[] prefix, int nTokens, int k, Rng rng) {
        if (k < 1) {
            throw new IllegalArgumentException("k must be >= 1");
        }
        Trace trace = new Trace();
        List<Integer> ctx = new ArrayList<>();
        for (int t : prefix) {
            ctx.add(t);
        }
        while (trace.tokens.size() < nTokens) {
            List<Integer> emitted = speculativeStep(target, draft, toArray(ctx), k, rng, trace);
            ctx.addAll(emitted);
        }
        while (trace.tokens.size() > nTokens) {
            trace.tokens.remove(trace.tokens.size() - 1);
        }
        return trace;
    }

    /**
     * Generates {@code nTokens} tokens by sampling the target model directly,
     * one forward pass per token. Used as the baseline.
     *
     * @param target the target model
     * @param prefix the initial context
     * @param nTokens the number of tokens to generate
     * @param rng the uniform source
     * @return the completed trace
     */
    public static Trace generateTargetOnly(Model target, int[] prefix, int nTokens, Rng rng) {
        Trace trace = new Trace();
        List<Integer> ctx = new ArrayList<>();
        for (int t : prefix) {
            ctx.add(t);
        }
        for (int i = 0; i < nTokens; i++) {
            double[] p = target.nextDist(toArray(ctx));
            int x = Distributions.sample(p, rng);
            trace.tokens.add(x);
            trace.targetPasses += 1;
            trace.steps += 1;
            ctx.add(x);
        }
        return trace;
    }

    /**
     * Returns the idealised expected speedup: emitted tokens per step divided by
     * the relative cost of one step (one target pass plus {@code k} draft passes
     * each costing {@code draftCostRatio}).
     *
     * @param tokensPerStep average tokens emitted per step
     * @param k the number of speculative proposals per step
     * @param draftCostRatio cost of a draft pass relative to a target pass
     * @return the expected speedup factor
     */
    public static double expectedSpeedup(double tokensPerStep, int k, double draftCostRatio) {
        return tokensPerStep / (1.0 + k * draftCostRatio);
    }
}
